pandas/tests/extension/base/__init__.py

```diff
@@ -41,26 +41,26 @@ class TestMyDtype(BaseDtypeTests):
 ``assert_series_equal`` on your base test class.
 
 """
-from .casting import BaseCastingTests  # noqa
-from .constructors import BaseConstructorsTests  # noqa
-from .dtype import BaseDtypeTests  # noqa
-from .getitem import BaseGetitemTests  # noqa
-from .groupby import BaseGroupbyTests  # noqa
-from .interface import BaseInterfaceTests  # noqa
-from .io import BaseParsingTests  # noqa
-from .methods import BaseMethodsTests  # noqa
-from .missing import BaseMissingTests  # noqa
-from .ops import (  # noqa
+from pandas.tests.extension.base.casting import BaseCastingTests  # noqa
+from pandas.tests.extension.base.constructors import BaseConstructorsTests  # noqa
```

Comment:
Personally, I am -1 on general linting check for this. For example, this specific example where I am commenting on, I find the relative imports actually more readable.

Reply:
I agree with Joris on this (more generally, i like relative imports for same-directory as it clarified dependency structure)

Reply:
I have no strong opinion here, I was just following the (existing) pandas style guide - should the style guide be changed so an exception is made for same-level imports?

Reply:
if we had easily automatic check able for relative imports that would be ok for a limited purpose

we cannot expreess 'more readable' in this way and have it enforced 

this is why i am currently -1 on relative imports and why we cannot use them in the code base

Reply:
> I was just following the (existing) pandas style guide - should the style guide be changed so an exception is made for same-level imports?

The part of the style guide about general relative imports, of course fully agreed that we never want implicit relative imports. But personally I think we should reconsider explicit ones, yes.

> more generally, i like relative imports for same-directory as it clarified dependency structure

+1, I also find this helpful. I think it should certainly be doable to have a check that relative imports are never outside the local directory

Reply:
At the moment it reads

> Explicit relative imports are also supported in Python 3 but it is not recommended to use them.

It certainly would be doable to customise the rule, I just don't (yet) see why you'd make an exception for same-directory imports. Doesn't `from ..foo import bar` also clarify the dependency structure? At which point, why not use explicit relative imports everywhere like sklearn does?

Reply:
>  I just don't (yet) see why you'd make an exception for same-directory imports. Doesn't from ..foo import bar also clarify the dependency structure?

there are a bunch of places where we have dependency structure rules (imperfectly enforced, so maybe better thought of as aspirations) of the form "this file doesnt depend on anything else in pandas except other files in this directory or things explicitly higher up in the dependency structure":

pandas._config
pandas._libs.tslibs (depends only on _config)
pandas._libs (depends almost-only on tslibs)
pandas.tseries (depends only on tslibs)
pandas.core.dtypes (depends almost-only on _libs)
pandas.core.arrays (depends almost^2-only on dtypes)
pandas.core.indexes (doesnt depend in groupby or internals)

Reply:
Thanks for explaining, I hadn't appreciated that - I'll look into this and see how to adapt imports

Reply:
> pandas.core.arrays (depends almost^2-only on dtypes)

@jbrockmendel just to make sure I've understood, are you suggesting that every file inside `pandas/core/arrays` from import from `pandas/core/arrays` in a relative manner? For example, in `pandas/core/arrays/sparse/array.py` there is
```
from pandas.core.arrays import ExtensionArray
```
, would you want that to be `from .. import ExtensionArray` instead?

Reply:
> would you want that to be from .. import ExtensionArray instead?

No, I would essentially never use `from .. import`, only `from . import`

Note that isort has a LOCALFOLDER section, so `from . import`s get put below everything else, which is consistent with the pattern of laying out imports in rough order of dependency structure

Reply:
OK, so you'd always use a relative import when importing from the same directory (even outside of the folders you listed above), and use absolute elsewhere? If so, this seems easily enforceable, I can make a PR later this week